GoalEntity 생성 후 Realm에 추가하는 기능 우선 구현

junu0516 opened the pull request:
내일 만나기 전에 GoalEntity를 Realm에 추가하는 부분을 같이 봐야할듯해서 우선 작업한 부분 먼저 PR 올렸어요!
크게 작업한건 없고, 기존에 구현해주셨던 CalendarView에서 선택한 데이터로 전달받은 Period 가지고 Goal -> GoalEntity 순서로 생성해서 Realm에 추가 후 확인하는 흐름입니다.
아직 진짜 추가만해서 미흡한 점이 많아서 ㅋㅋ 같이 한 번 보고 피드백 주셔서 다시 역할 분담해서 부족한 부분 더 구현하고 리팩토링하면 될 것 같아요!

junu0516 committed:
feat: [#61] InsertField + Reactive 추가, GoalAddRepository 구현 및 추가

- InsertField 내부의 textField의 값 변화 관찰을 위해 ControlProperty 타입 프로퍼티를 추가

junu0516 committed:
feat: [#61] InsertField 내부의 textField의 값 변화를 관찰하기 위한 속성 및 메소드 추가

- 우선 textField의 값이 변할 때마다 textFieldDidChange 함수를 호출해서, 내부의 text 속성에 textField의 값을 저장
- 이후 text 의 값이 변할 때 마다 didSet을 통해 sendActions로 액션을 방출
- .valueChanged 액션을 방출하기 때문에 외부에서는 rx 연산자로 Observable한 insertField의 text값 변화를 관찰할 수 있음

junu0516 committed:
refactor: [#61] goalAddButton을 터치했을 때 어느 카테고리인 지 알 수 있도록 Category 타입을 같이 넘겨주도록 수정

junu0516 committed:
refactor: [#61] GoalCoordinator의 showGoalAddView의 제네릭 타입을 Category으로 수정

junu0516 committed:
feat: [#61] GoalAddViewController에서 titleField, descriptionField의 text값에 대한 바인딩 추가

junu0516 committed:
refactor: [#61] Goal, GoalEntity 타입에 Period 관련 타입 속성 선언 추가

junu0516 committed:
feat: [#61] 저장 버튼 눌렀을 때 새로운 Goal, GoalEntity 추가 후 리스트 업데이트 확인

junu0516 commented on rabit/rabit/Utilities/InsertField + Reactive.swift:
```diff
@@ -0,0 +1,16 @@
+import RxSwift
+import RxCocoa
+
+extension Reactive where Base == InsertField {
+    
+    var text: ControlProperty<String> {
+        base.rx.controlProperty(
+            editingEvents: .valueChanged,
+            getter: { $0.text ?? "" },
+            setter: { insertField, text in
+                insertField.text = text
+            }
+        )
+    }
+}
```
InsertField라는 UIView를 상속받은 클래스 내부에서 가지고 있는 UITextField 속성이 가지고 있는 텍스트 입력값을 전달받기 위해서 ControlProperty를 추가로 또 선언했어요!

다만 이 코드 작성하면서 생각든게, InsertField라는 컴포넌트뷰 자체가 디자인도 미흡하고 기능상으로도 미흡한 부분이 많아서 추후에 InsertField 자체를 해체하고 이것이 가지고 있던 뷰 속성들을 GoalAddViewController이 직접 가지고 있는 식으로 변경할 수도 있겠다 싶었어요!

이럴 경우에는 해당 코드는 삭제해도 무방할듯 합니다 ㅎㅎ

Hansolkkim replied:
어제 말씀하셨던 부분이군요ㅋㅋ 이건 그냥 제 의견인데, 그냥 InsertField를 뷰를 묶어놓은 컨테이너 정도로 생각하고, InsertField 내에 있는 textField를 private하지 않게 선언하시면 따로 값 설정이나 입력된 값 받아오는걸 이렇게 extension으로 안해도 괜찮지 않을까 생각이 듭니다!

junu0516 replied:
외부에서 접근 가능하게 하는게 아무래도 더 나아보이죠?ㅋㅋㅋ
그렇게 바꿔서 불필요한 코드 없애야겠어요 ㅎㅎ 숨기기 위해서 처리해야될게 더 많아지다 보니까 배보다 배꼽이 더 커지는 것 같은..

junu0516 commented on rabit/rabit/Goal/GoalAdd/Views/InsertField.swift:
```diff
@@ -1,7 +1,7 @@
 import SnapKit
 
-final class InsertField: UIView {
+final class InsertField: UIControl {
```
커스텀하게 ControlProperty 속성을 추가선언하려면 UIView가 아닌 UIControl을 상속받아야 해서 위와 같이 바꿨어요!

junu0516 commented on rabit/rabit/Goal/GoalAdd/Views/InsertField.swift:
```diff
@@ -65,4 +67,8 @@ final class InsertField: UIView {
             $0.height.equalToSuperview().multipliedBy(0.5)
         }
     }
+    
+    @objc private func textFieldDidChange(_ textField: UITextField) {
+        text = textField.text
+    }
 }
```
InsertField가 가지고 있는 UITextField 속성의 text값이 변할 때마다 자체적으로 가진 text 속성에 변한 값을 그대로 전달시켜주도록 메소드를 추가했어요!
최종적으로 외부에서는 InsertField의 text 속성값의 변화를 바라보는 식으로 간접적으로 UITextField 속성의 입력값 변화를 관찰하게 됩니다!

junu0516 commented on rabit/rabit/Goal/GoalAdd/GoalAddViewModel.swift:
```diff
@@ -59,5 +64,43 @@ private extension GoalAddViewModel {
             .bind(to: navigation.showTimeSelectView)
             .disposed(by: disposeBag)
+        
+        let goal = Observable
+                    .combineLatest(
+                        goalTitleInput,
+                        goalSubtitleInput,
+                        selectedPeriod,
+                        selectedTime
+                    )
+                    .withUnretained(self)
+                    .map {
+                        Goal(
+                            title: $1.0,
+                            subtitle: $1.1,
+                            progress: .zero,
+                            period: $1.2,
+                            certTime: $1.3,
+                            target: .zero,
+                            category: $0.category.title
+                        )
+                    }
+                    .share()
```
이부분 급하게 Goal 데이터가 잘 추가되나 보려고 선언한 지역변수입니다 ㅋㅋ
map 함수를 사용하면서 뭔가 코드가 좀 길어졌는데 전에 알려주신대로 좀 더 간단한 문법으로 축약할 수 있었던 것 같은데 갑자기 생각이 안나서;; 우선 커밋했어요!
아마 수정해야할듯 합니다 ㅎㅎ

junu0516 committed:
feat: [#61] Date에 Strideable 채택 추가

junu0516 committed:
feat: [#61] Goal의 target 속성을 연산프로퍼티로 변경

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/Views/InsertField.swift:
```diff
@@ -38,9 +40,9 @@ final class InsertField: UIView {
     
-    var isEnabled: Bool = true {
+    var isTextFieldEnabled: Bool = true {
         didSet {
-            textField.isEnabled = isEnabled
+            textField.isEnabled = isTextFieldEnabled
         }
```
이 프로퍼티는 나중에 Goal Cell을 터치했을때 따로 수정은 못 하도록 할때 사용하려고 존재하는 프로퍼티인가요?

junu0516 replied:
넵 원래는 isEnabled로 이름지었는데, UIControl을 상속받는 형태로 바꾸면서 UIControl이 기존에 가지고 있던 isEnabled랑 겹쳐서 이름을 바꿨어요

Hansolkkim commented on rabit/rabit/Goal/GoalCoordinator.swift:
```diff
@@ -6,7 +6,7 @@ protocol GoalNavigation {
     var showCategoryAddView: PublishRelay<Void> { get }
     var closeCategoryAddView: PublishRelay<Void> { get }
-    var showGoalAddView: PublishRelay<Void> { get }
+    var showGoalAddView: PublishRelay<Category> { get }
```
이 부분은 Goal을 새로 추가할때 Category Title을 알고 있어야해서 추가하신거죠?!

junu0516 replied:
넵 맞습니다 ㅎㅎ

Hansolkkim commented on rabit/rabit/Goal/Models/Goal.swift:
```diff
@@ -5,20 +5,31 @@ struct Goal {
-    let target: Int
     let category: String
+    
+    var target: Int {
+        var count = 0
+        let days = certTime.days.selectedValues
+        for date in stride(from: period.start, to: period.end, by: 60*60*24) {
+            guard let day = Day(rawValue: Calendar.current.component(.weekday, from: date) - 1) else {
+                continue
+            }
+            if days.contains(day) { count += 1 }
+        }
+        return count
+    }
```
일수 계산을 위한 로직이군요..ㅋㅋ 대충봐도 복잡해보입니다..ㅋㅋㅋ
혹시 guard 구문은 어떤걸 확인하는 부분인가요?

junu0516 replied:
Day가 열거형인데, `rawValue` 를 사용하는 생성자가 실패가능한 생성자더라구요,, 그래서 옵셔널바인딩 해주려고 사용했어요!

Hansolkkim commented on rabit/rabit/Goal/Models/GoalEntity.swift:
```diff
@@ -7,6 +8,8 @@ final class GoalEntity: Object {
     @Persisted var progress: Int = 0
     @Persisted var target: Int = 0
     @Persisted var category: String = ""
+    @Persisted var startDate: Date = Date()
+    @Persisted var endDate: Date = Date()
```
이 부분은 Realm에 Period 타입을 저장할 수 없어서 이렇게 분리하신거 맞을까요?

junu0516 replied:
넵넵 Period에 대한 Entity를 따로 정의하거나 다른 방식으로 커스텀 타입을 만들어도 되는데, 좀 복잡할듯 싶어서 그냥 Period가 가지고 있던 Date 타입 속성을 바깥으로 꺼냈어요!

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/GoalAddViewModel.swift:
```diff
@@ -55,5 +64,42 @@ private extension GoalAddViewModel {
             .bind(to: navigation.showTimeSelectView)
             .disposed(by: disposeBag)
+        
+        let goal = Observable
+                    .combineLatest(
+                        goalTitleInput,
+                        goalSubtitleInput,
+                        selectedPeriod,
+                        selectedTime
+                    )
+                    .withUnretained(self)
+                    .map {
+                        Goal(
+                            title: $1.0,
+                            subtitle: $1.1,
+                            progress: .zero,
+                            period: $1.2,
+                            certTime: $1.3,
+                            category: $0.category.title
+                        )
+                    }
+                    .share()
```
```suggestion
        let goal = Observable
            .combineLatest(
                  goalTitleInput, 
                  goalSubtitleInput, 
                  selectedPeriod, 
                  selectedTime
             )
            .withUnretained(self) { ($1.0, $1.1, .zero, $1.2, $1.3, $0.category.title) }
            .map(Goal.init(title:subtitle:progress:certTime:target:category:))
            .share()
```

이거 말씀하신거 같네요! 그런데 생성자의 `progress: .zero`는 어떤 의미인가요? progress는 Int 타입인데 .zero가 가능한건지 궁금합니다!

그리고 share()를 쓰든 안쓰든 `goal`는 Observable<Goal> 타입인거 같은데, 혹시 어떤 이유로 붙이는건지 궁금합니다!ㅎㅎ 이건 개인적인 질문입니다..ㅎㅎ

Hansolkkim replied:
suggested change 부분에 `.zero`는 해당 값이 어떤 의미를 가지는지 몰라서 일단 대충 작성한거라 `commit suggestion` 바로 하시면 안될거에요 아마도!

junu0516 replied:
`.zero` 는 맨 처음 생성할 때 progress는 값이 없는 상태여야 하기 때문에 사용했어요! target = 분모, progress = 분자 이렇게 보면 분자의 초기값을 0으로 둔 셈으로 보시면 됩니다!

junu0516 replied:
`share()` 같은 경우는 위에서 Goal 타입 인스턴스를 생성하는 Observable 시퀀스는 구독할 때마다 Observable을 생성하는 형태인데, 이렇게 되면 여러 곳에서 해당 시퀀스를 구독할 때마다 불필요하게 Observable이 너무 많이 생성되기 때문에 이걸 방지하려고 사용했어요!
![image](https://user-images.githubusercontent.com/68586291/194279645-6fc19c0a-24ee-4cbc-98af-858acdbd0d46.png)

`share()` 을 사용하면 한 번 생성된 시퀀스를 여러 곳에서 공유할 수 있는 형태이기 때문에, 이후의 여러 구독 객체들은 해당 시퀀스를 공유하게 됩니다! 
(사실 여기서는 크게 의미없는 사용이긴 한데... 저도 RxSwift 학습하다가 저렇게 `share()` 을 사용한걸 보고 궁금해서 이번에 사용해봤어요 ㅋㅋ)
![image](https://user-images.githubusercontent.com/68586291/194279699-7850579c-5c6e-4280-af59-e411f0dd60dc.png)

junu0516 replied:
@Hansolkkim 
말나온김에 ㅋㅋ
이전까지는 잘 몰라서.. RxSwift 오퍼레이터 사용한 코드는 흐름만 이래하는 식이었는데, `share()` 사용 보고 Observable 동작원리 한 번 더 공부하다보니까 이런 식으로 불필요하게 시퀀스가 생성되는 흐름이 있었겠구나(?)하는 생각도 들었어요 ㅋㅋ 필요한 기능 우선구현 끝나면 이 부분 한번 더 쭉 점검해도 좋을 것 같아요!

Hansolkkim replied:
오 자세한 설명 감사합니다.. 저도 이렇게 사용하려고 만드는 Onservable들이 많은데 share()로 처리하는게 더 맞겠네요!!

Hansolkkim reviewed (approved):
고생하셨습니다!
변경이 꽤 많았네요 이번에도ㅎㅎ
리뷰라기 보다는 모르는 부분 질문 위주로 코멘트 작성했습니다!

junu0516 committed:
feat: [#61] Realm에서 Goal 데이터 가져올 때, Photo의 개수 세서 progress 속성 업데이트

junu0516 committed:
refactor: [#61] Goal 안의 target 구하는 연산프로퍼티를 propertyWrapper 사용으로 대체

- 기존에는 연산프로퍼티를 사용했기 때문에, 컬렉션뷰에서 셀을 그릴 때마다 매번 불필요하게 여러 번 연산을 수행해야 하는 상황이었음
- 따라서 target 값을 초기화하는 연산은 Goal 인스턴스 생성 시 최초 1회만 수행되도록 해야 함
- 단 Goal 안에 값을 초기화하는 연산을 포함시키면 책임이 너무 커지기 때문에 propertyWrapper을 사용해서 별도의 객체가 연산을 수행하도록 분리
